modules/sdk-lib-mpc/src/tss/ecdsa/noSmallFactorsProof.ts

```diff
@@ -0,0 +1,145 @@
+/**
+ * Implementation of No Small Factors ($\Pi^\text{fac}).
+ * https://eprint.iacr.org/2020/492.pdf Section B.4
+ */
+import { createHash } from 'crypto';
+import { bitLength, randBetween } from 'bigint-crypto-utils';
+import { modPow } from 'bigint-mod-arith';
+import { bigIntFromBufferBE, bigIntToBufferBE } from '../../util';
+import { DeserializedNoSmallFactorsProof } from './types';
+
+const ORDER = BigInt('0xfffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141');
+const ELL = BigInt(256);
+const EPSILON = BigInt(BigInt(2) * ELL);
+
+/**
+ * Generate pseudo-random challenge value $e$ for $(N, w)$.
+ * @param N - the prime number to verify is a product of two large primes.
+ * @param w - a random number with the same bitLength as N, that satisfies the Jacobi of w is -1 wrt N.
+ * @returns {bigint} - challenge value $e$.
+ */
+function generateE(N: bigint, w: bigint): bigint {
+  const digest = createHash('shake256', { outputLength: 1 + Math.floor((bitLength(ORDER) + 7) / 8) })
+    .update(bigIntToBufferBE(N))
+    .update('$')
+    .update(bigIntToBufferBE(w))
+    .update('$')
+    .digest();
+  const e = bigIntFromBufferBE(digest.subarray(1)) % ORDER;
```

Comment:
This has the same issue as BIG-13-003 from Cure53's audit, where the random sampling of the challenge is not uniform from [0, order).  Can be done in a follow up task though.